### PR TITLE
Enforce unwind invariants

### LINKED ISSUE #72959: -Zvalidate-mir: check that all unwind edges go to `cleanup` blocks
Blocks that are part of unwinding code always need to marked as `cleanup`. This seems like an easy thing to check in the validation pass. I believe this would have caught a bug in the generator transform (a pretty long time ago).

This issue has been assigned to @doctorn via [this comment](https://github.com/rust-lang/rust/issues/72959#issuecomment-640562605).

### PR DESCRIPTION
**Pull request by cofibrant**
I had a quick look at #72959. The failure message probably needs to be more detailed but I just wanted to check I got the right idea. I have no idea how to right a test for this either...

r? @jonas-schievink 

Resolves #72959 (hypothetically)

### CODE REVIEW

File: src/librustc_mir/transform/validate.rs [hunk 1]
```diff
@@ -49,8 +54,25 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
         );
     }
 
-    fn check_bb(&self, location: Location, bb: BasicBlock) {
-        if self.body.basic_blocks().get(bb).is_none() {
+    fn check_bb(&self, location: Location, bb: BasicBlock, edge_kind: EdgeKind) {
```
Comment (jonas-schievink) [inline src/librustc_mir/transform/validate.rs:57]: Probably a better name for this function would now be `check_edge`

File: src/librustc_mir/transform/validate.rs [hunk 2]
```diff
@@ -49,8 +54,25 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
...
+                (false, false, EdgeKind::Normal)
+                // Non-cleanup blocks can jump to cleanup blocks along unwind edges
+                | (false, true, EdgeKind::Unwind)
+                // Cleanup blocks can jump to cleanup blocks along non-unwind edges
+                | (true, true, EdgeKind::Normal) => {}
+                // All other jumps are invalid
+                _ => {
+                    self.fail(
+                        location,
+                        format!("encountered jump that does not respect unwind invariants {:?}", bb)
```
Comment (jonas-schievink) [inline src/librustc_mir/transform/validate.rs:71]:
```suggestion
                        format!("{:?} edge to {:?} violates unwind invariants (cleanup {:?} -> {:?})", edge_kind, bb, src.is_cleanup, bb.is_cleanup)
```

**Review by jonas-schievink — COMMENTED** [review]
Looks pretty good! I'll run this with `-Zvalidate-mir` on the test suite to check if it finds anything. I don't think we need more testing than that (this is only an internal utility pass after all).

**Comment by jonas-schievink** [comment]
> I'll run this with `-Zvalidate-mir` on the test suite to check if it finds anything.

Oh, right, this doesn't currently work

File: src/librustc_mir/transform/validate.rs (hunk 1 above)
Comment (doctorn) [reply src/librustc_mir/transform/validate.rs:57]: I agree - I’ll update tomorrow

File: src/librustc_mir/transform/validate.rs (hunk 2 above)
Comment (doctorn) [reply src/librustc_mir/transform/validate.rs:71]: This is a great suggestion thanks a lot

**Pull request merged**